packages/flutter_tools/lib/src/android/gradle_utils.dart

```diff
@@ -145,66 +168,407 @@ String getGradleVersionForAndroidPlugin(Directory directory, Logger logger) {
-      parsedTargetVersion <= maxVersion;
+/// Returns either the gradle-wrapper.properties value from the passed in
+/// [directory] or if not present the version available in local path.
+///
+/// If gradle version is not found null is returned.
+/// [directory] should be and android directory with a build.gradle file.
+Future<String?> getGradleVersion(
+    Directory directory, Logger logger, ProcessManager processManager) async {
+  final File propertiesFile = directory
+      .childDirectory(gradleDirectoryName)
+      .childDirectory(gradleWrapperDirectoryName)
+      .childFile(gradleWrapperPropertiesFilename);
+
+  if (propertiesFile.existsSync()) {
+    final String wrapperFileContent = propertiesFile.readAsStringSync();
+
+    // Expected content format (with lines above and below).
+    // Version can have 2 or 3 numbers.
+    // 'distributionUrl=https\://services.gradle.org/distributions/gradle-7.4.2-all.zip'
+    final RegExp distributionUrlRegex = RegExp(r'distributionUrl=.*\.zip');
+
+    final RegExpMatch? distributionUrl =
+        distributionUrlRegex.firstMatch(wrapperFileContent);
+    if (distributionUrl != null) {
+      // Expected content: 'gradle-7.4.2-all.zip'
+      final String? gradleZip = distributionUrl.group(0);
+      if (gradleZip != null) {
+        final String gradleVersion = gradleZip.split('-')[1];
+        return gradleVersion;
+      } else {
+        // Did not find gradle zip url. Likley this is a bug in our parsing.
+        logger.printWarning(_formatParseWarning(wrapperFileContent));
+      }
+    } else {
+      // If no distributionUrl log then treat as if there was no propertiesFile.
+      logger.printTrace(
+          '$propertiesFile does not provide an Gradle version falling back to system gradle.');
+    }
+  } else {
+    // Could not find properties file.
+    logger.printTrace(
+        '$propertiesFile does not exist falling back to system gradle');
+  }
+  // System installed Gradle version.
+  if (processManager.canRun('gradle')) {
+    final String gradleVersionVerbose =
+        (await processManager.run(<String>['gradle', gradleVersionFlag])).stdout
+            as String;
+    // Expected format:
+/*
+
+------------------------------------------------------------
+Gradle 7.6
+------------------------------------------------------------
+
+Build time:   2022-11-25 13:35:10 UTC
+Revision:     daece9dbc5b79370cc8e4fd6fe4b2cd400e150a8
+
+Kotlin:       1.7.10
+Groovy:       3.0.13
+Ant:          Apache Ant(TM) version 1.10.11 compiled on July 10 2021
+JVM:          17.0.6 (Homebrew 17.0.6+0)
+OS:           Mac OS X 13.2.1 aarch64
+    */
+    // Observation shows that the version can have 2 or 3 numbers.
+    // Inner parentheticals `(\.\d+)?` denote the optional third value.
+    // Outter parentheticals `Gradle (...)` denote a grouping used to extract
+    // the version number.
+    final RegExp gradleVersionRegex = RegExp(r'Gradle (\d+\.\d+(\.\d+)?)');
```

Comment:
The inner group here can be non-capturing.

Reply:
Done.